libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/InternalUpsert.cs

```diff
@@ -74,6 +74,8 @@ internal OperationStatus InternalUpsert<TInput, TOutput, TContext, TSessionFunct
                 if (stackCtx.recSrc.HasReadCacheSrc)
                 {
                     srcRecordInfo = ref stackCtx.recSrc.GetInfo();
+                    // we are about to do an upsert so even if the prev data had etag we need to reset it at upsert
+                    srcRecordInfo.ClearHasETag();
                     goto CreateNewRecord;
```

Comment:
Don't clear this. If we fail the CAS we will unlock for the retry and someone can read it then and won't see the ETag. Same below. Instead just make sure your ConcurrentWriter and SingleWriter leave it unset